Reintroduce test suite for install script

ptgott opened the pull request:
The original location of the Teleport installation script included a test suite based on Docker Compose. Add the test suite to the new location along with the original GitHub Actions workflow that ran on changes to the installation script.

ptgott committed:
Reintroduce test suite for install script

The original location of the Teleport installation script included a
test suite based on Docker Compose. Add the test suite to the new
location along with the original GitHub Actions workflow that ran on
changes to the installation script.

camscale commented on .github/workflows/test-installation.yaml:
Can you rename this workflow file to `test-install-script` or similar? `test-installation` sounds like it is testing some installation of teleport.

camscale commented on assets/install-scripts/install-teleport-tests/run-test.sh:
```diff
@@ -0,0 +1,60 @@
+
+function test_upgrader() {
+    if ! type teleport-upgrade; then
+    	echo "INSTALL_SCRIPT_TEST_FAILURE: upgrader not found"
```
s/upgrader/teleport-upgrade/ - it is helpful if error messages are specific.

camscale commented on assets/install-scripts/install-teleport-tests/run-test.sh:
```diff
@@ -0,0 +1,60 @@
+echo "RUNNING TEST"
+
+case $1 in
+    cloud)
+    	test_teleport
+    	test_tctl
+    	test_tsh
+    	test_tbot
+    	test_upgrader
+    ;;
+    oss)
+    	test_teleport
+    	test_tctl
+    	test_tsh
+    	test_tbot
+    ;;
+    enterprise)
+    	test_teleport
+    	test_tctl
+    	test_tsh
+    	test_tbot
+    ;;
```
If you want, you could use the `;;&` case terminator to keep evaluating cases and re-write these cases as:
```
    cloud)
      test_upgrader
      ;;&
    oss | enterprise | cloud)
      test_teleport
      test_tctl
      test_tsh
      test_tbot
      ;;

```
The `;;&` causes bash to continue evaluating cases.

camscale commented on assets/install-scripts/install-teleport-tests/run-test.sh:
```diff
@@ -0,0 +1,60 @@
+    	test_tbot
+    ;;
+    *)
+    	echo "INSTALL_SCRIPT_TEST_FAILURE: unsupported edition $EDITION"
```
`EDITION` is not set anywhere in this script.

camscale commented on assets/install-scripts/install-teleport-tests/docker-compose.yml:
```diff
@@ -0,0 +1,102 @@
+version: "3.8"
```
When I run docker compose I get an error: `the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion` (docker compose 2.29.6)

camscale commented on assets/install-scripts/install-teleport-tests/run-test.sh:
```diff
@@ -0,0 +1,60 @@
+    	test_teleport
+    	test_tctl
+    	test_tsh
+    	test_tbot
```
should we also test that `fdpass-teleport` is installed? That's probably a later version than 15.0.0 but I note the install script does mention it if it exists.

Although now I think on it more, I think it might be better to remove the tests for `tctl`, `tsh` and `tbot` - the installed does nothing specific to install those - it just installs a teleport release asset so we should only be checking that it has done that, not that the release asset contains the correct binaries - that test belongs somewhere else in the release process perhaps. Testing for the upgrader for cloud is correct as the install script has special logic for that.

ptgott replied:
Ah, good point. Changed.

camscale commented on assets/install-scripts/install-teleport-tests/README.md:
````diff
@@ -0,0 +1,66 @@
+1. Edit the `command` field of the service definition to include the following:
+
+   ```yaml
+          bash /install.sh 15.0.0;
````
I notice that some of the tests have 15.1.0 instead of 15.0.0 - what determines the version that goes here?

ptgott replied:
To be honest, I don't think I was thinking about this. When I first added the test environment in the private repo, 15 was the latest major version, and beyond that, I think it was pretty arbitrary.

What would be the most maintainable way to set a version here? Have `run-all-tests.sh` retrieve the latest release and pass it to the `install.sh` script using a container environment variable?

camscale replied:
That sounds reasonable and removes a maintenance burden on managing the multiple version numbers currently here.

camscale commented on assets/install-scripts/install-teleport-tests/docker-compose.yml:
This file has a `.yml` extension while the workflow file has a `.yaml` extension. We should probably stick to the latter since we are not constrained by DOS 3 letter extensions (although I note we do have a mixture of these extensions in our repo, but two different ones in the same PR I think is pushing it :) )

ptgott commented:
TODO Review feedback to incorporate:
- [ ] https://github.com/gravitational/teleport/pull/46718#discussion_r1767990403
- [ ] https://github.com/gravitational/teleport/pull/46718#discussion_r1767992870
- [ ] https://github.com/gravitational/teleport/pull/46718#discussion_r1767996752
- [ ] https://github.com/gravitational/teleport/pull/46718#discussion_r1767997011
- [ ] https://github.com/gravitational/teleport/pull/46718#discussion_r1768003222
- [ ] https://github.com/gravitational/teleport/pull/46718#discussion_r1768007618
- [ ] https://github.com/gravitational/teleport/pull/46718#discussion_r1768008722
- [ ] https://github.com/gravitational/teleport/pull/46718#discussion_r1768011271

ptgott committed:
Partially respond to camscale feedback

- Rename the workflow
- Fix an error message
- Fix unused variable
- docker-compose.yml -> docker-compose.yaml
- Remove obsolete `version` attribute from docker-compose.yaml
- Refactor case statement in run-test.sh
- Remove unnecessary test assertions. The test doesn't need to check
  that all expected binaries are in a teleport package.

ptgott commented:
@camscale Following up to see if I addressed your feedback. Thanks!

camscale commented on assets/install-scripts/install-teleport-tests/docker-compose.yaml:
```diff
@@ -0,0 +1,101 @@
+       bash -c 'apt-get update;
+       apt-get install -y curl;
+       sed -E -i "s/ID_LIKE=.*$/ID_LIKE=\"ubuntu debian\"/" /etc/os-release;
+       bash /install.sh 15.0.0;
```
nit: for consistency with the rest of this file, this should have `oss` as an arg to the install script (I know it's the default, just that the other oss install has it explicitly)

camscale reviewed (approved):
I think this is fine as is, but would be better with the auto-versioning. That could be a follow-up PR if you prefer.

ptgott committed:
Use latest release version in install script tests

This way, we can guarantee that the release version we are installing in
tests available. Use `gh` to get the latest release number and pass the
TELEPORT_VERSION shell variable to test containers.

ptgott committed:
Edit the installation script

Get tests to pass by not specifying a Teleport version if using apt-get
with Teleport Cloud. This is because the version specified by the
`TELEPORT_VERSION` variable may not be available in the `stable/cloud`
channel.

Also remove redundant `/etc/os-release` sourcing.

ptgott commented:
Going to close this one as it's out of date with the way we currently handle the install script (e.g., it's now in `web/scripts/install/install.sh`) and I want to reduce the number of in-flight projects I have.
